File: problems/tier1-foundation/003-notification-system/tests/cpp/part1_test.cpp
// Part 1 Tests — Notification System
// Tests basic observer pattern: subscribe, notify, multi-channel

#include "solution.cpp"
#include <cassert>
#include <iostream>
#include <vector>
#include <string>
using namespace std;

// Track what was sent (tests inject a spy observer)
vector<string> sent_notifications;

int part1_tests() {
    int passed = 0;
    int failed = 0;
    sent_notifications.clear();

    // Test 1: notify sends to subscribed channels
    try {
        User u1 = {"user1", "u1@test.com", "+91-9000000001", {"email"}};
        User u2 = {"user2", "u2@test.com", "+91-9000000002", {"sms", "push"}};
        User u3 = {"user3", "u3@test.com", "+91-9000000003", {"email", "sms"}};

        // notify should not throw
        notify("Order shipped", {u1, u2, u3});
        cout << "PASS test_notify_no_throw" << endl;
        passed++;
    } catch (...) {
        cout << "FAIL test_notify_no_throw" << endl;
        failed++;
    }

    // Test 2: user not subscribed to a channel receives nothing on that channel
    try {
        // If we can capture output, verify only subscribed channels receive
        // For simplicity, we test that an empty subscriber list causes no crash
        vector<User> empty;
        notify("Event", empty);
        cout << "PASS test_empty_user_list" << endl;
        passed++;
    } catch (...) {
        cout << "FAIL test_empty_user_list" << endl;
        failed++;
    }

    // Test 3: user subscribed to multiple channels receives on each
    try {
        User u = {"u1", "u1@test.com", "+1-555-0001", {"email", "sms", "push"}};
        // Should not throw even with multiple channels
        notify("Flash sale", {u});
        cout << "PASS test_multi_channel_user" << endl;
        passed++;
    } catch (...) {
        cout << "FAIL test_multi_channel_user" << endl;
        failed++;
    }

    cout << "PART1_SUMMARY " << passed << "/" << (passed + failed) << endl;
    return failed;
}

File: problems/tier1-foundation/003-notification-system/tests/java/Part1Test.java
// Notification System — Part 1 Tests
import java.util.*;
import java.util.stream.*;

class Part1Test {
    static boolean testNotifyNoThrow() {
        try {
            User u1 = Arrays.asList("user1", "u1@test.com", "+91-9000000001", {"email")};
            User u2 = Arrays.asList("user2", "u2@test.com", "+91-9000000002", {"sms", "push")};
            User u3 = Arrays.asList("user3", "u3@test.com", "+91-9000000003", {"email", "sms")};
            // notify should not throw
            notify("Order shipped", {u1, u2, u3});
            boolean pass = true; // FIXME: no assertions found
            System.out.println((pass ? "PASS" : "FAIL") + ": testNotifyNoThrow");
            return pass;
        } catch (Exception e) {
            System.out.println("FAIL: testNotifyNoThrow (exception: " + e.getMessage() + ")");
            return false;
        }
    }

    static boolean testEmptyUserList() {
        try {
            // If we can capture output, verify only subscribed channels receive
            // For simplicity, we test that an empty subscriber list causes no crash
            List<User> empty;
            notify("Event", empty);
            boolean pass = true; // FIXME: no assertions found
            System.out.println((pass ? "PASS" : "FAIL") + ": testEmptyUserList");
            return pass;
        } catch (Exception e) {
            System.out.println("FAIL: testEmptyUserList (exception: " + e.getMessage() + ")");
            return false;
        }
    }

    static boolean testMultiChannelUser() {
        try {
            User u = Arrays.asList("u1", "u1@test.com", "+1-555-0001", {"email", "sms", "push")};
            // Should not throw even with multiple channels
            notify("Flash sale", {u});
            boolean pass = true; // FIXME: no assertions found
            System.out.println((pass ? "PASS" : "FAIL") + ": testMultiChannelUser");
            return pass;
        } catch (Exception e) {
            System.out.println("FAIL: testMultiChannelUser (exception: " + e.getMessage() + ")");
            return false;
        }
    }

    public static int runTests() {
        int passed = 0, total = 0;
        total++; if (testNotifyNoThrow()) passed++;
        total++; if (testEmptyUserList()) passed++;
        total++; if (testMultiChannelUser()) passed++;
        System.out.println("PART1_SUMMARY " + passed + "/" + total);
        return passed;
    }
}
